### PR TITLE
Add DMA support to the QEMU FW_CFG device

### LINKED ISSUE #3631: Add DMA support to the QEMU fw_cfg device driver
We use the fw_cfg device to read the initial RAM disk. The current implementation only supports reading one byte at a time. This means that loading of a 1MiB RAM disk takes 33 second. Presumable workloads will be 10s of megabytes, which means this would not be an acceptable solution from a performance point of view.

### PR DESCRIPTION
**Pull request by conradgrobler**
This allows us to load the initial RAM disk via DMA rather than 1 byte at a time.

Fixes #3631

### CODE REVIEW

File: stage0/src/fw_cfg.rs [hunk 1]
```diff
@@ -225,26 +233,25 @@ impl FwCfg {
     }
 
     /// Reads the address for the intended location or the initial RAM disk.
-    pub fn read_initrd_address(&mut self) -> Result<VirtAddr, &'static str> {
+    pub fn read_initrd_address(&mut self) -> Result<PhysAddr, &'static str> {
         let mut initrd_addr: u32 = 0;
         self.write_selector(FwCfgItems::InitrdAddr as u16)?;
         self.read(&mut initrd_addr)?;
         // Since we use an identity mapping in the Stage0 firmware, we can interpret the physical
```
Comment (andrisaar) [inline stage0/src/fw_cfg.rs:240]: This comment is now out of date, as we're not returning a VirtAddr.

File: stage0/src/fw_cfg.rs [hunk 2]
```diff
@@ -225,26 +233,25 @@ impl FwCfg {
...
         // address directly as a virtual address.
-        Ok(VirtAddr::new(initrd_addr as u64))
+        Ok(PhysAddr::new(initrd_addr as u64))
     }
 
     /// Reads the contents of the initial RAM disk into the supplied buffer.
     ///
     /// The initial RAM disk is not available via the file interface but has its own selector.
-    ///
-    /// Important note: we cannot read the size of the initial RAM disk again, as it returns 0 on
```
Comment (andrisaar) [inline stage0/src/fw_cfg.rs:241]: Side note, but I wonder if that's true for other fw_cfg entries as well?

File: stage0/src/fw_cfg.rs [hunk 3]
```diff
@@ -276,4 +283,65 @@ impl FwCfg {
...
+        /// Indicates that an error occurred during the DMA operation.
+        const ERROR = 1 << 0;
+        /// A DMA guest read operation is requested.
+        const READ = 1 << 1;
+    }
+}
+
+/// Definition for a DMA access request.
+#[repr(C)]
+struct FwCfgDmaAccess {
```
Comment (andrisaar) [inline stage0/src/fw_cfg.rs:332]: does this struct need any particular alignment?

**Review by andrisaar — APPROVED** [review]
Add a note somewhere (maybe file a bug so that we don't forget?) that this is not currently compatible with SEV-SNP.

File: stage0/src/fw_cfg.rs (hunk 1 above)
Comment (conradgrobler) [reply stage0/src/fw_cfg.rs]: Thanks for catching it. Done.

**Comment by conradgrobler** [comment]
> Add a note somewhere (maybe file a bug so that we don't forget?) that this is not currently compatible with SEV-SNP.

Create #3636 to track the SEV-SNP enlightenment.

**Pull request merged**